projects/sel4/porting.md

```diff
@@ -7,79 +7,159 @@ SPDX-FileCopyrightText: 2020 seL4 Project a Series of LF Projects, LLC.
+
+The following command allows you to decompile the DTB:
+```sh
+dtc -I dtb -O dts /path/to/dtb -o linux.dts
+```
+
+It should be noted that you may need to add device nodes depending on what Linux uses
+and what is required by seL4. For example, in the port of the ROCKPro64 a memory node
+and an extra timer node were needed so these had to be defined manually.
 
 ### Hardware generation script
 
-Depending on your platform and the way the DTS is defined you may need to modify the hardware_gen.py script.
-From the above example the Rockpro64 needed a memory node so that this script could define working memory.
-Other gotchas could include interrupt cells, interrupt numbers being mapped wrongly, or incorrect addresses.
-Also serial and timer drivers will need to be added to the kernel if they don't already exist. You can check this through
-the DTS device strings and compare to the hardware.yml file in `kernel/tools/hardware.yml`. If needed add the required drivers in `kernel/drivers/`.
+Depending on your platform and the way the DTS is defined you may need to modify the
+`seL4/tools/hardware_gen.py` script.
+
+From the above example the ROCKPro64 needed a memory node so that this script could
+define working memory. Other gotchas could include interrupt cells, interrupt numbers being
+mapped wrongly, or incorrect addresses. Also serial and timer drivers will need to be added
+to the kernel if they don't already exist. You can check this through the DTS device strings
+and compare to the `seL4/tools/hardware.yml` file.
+
+If needed, add the required drivers in `seL4/src/drivers/` and `seL4/include/drivers/`.
 
 ### Kernel
 
-You will need to add `libsel4/sel4_plat_include/<platform>/sel4/plat/api/constants.h` and add sel4_UserTop depending on what you support. Look in the other platform files for examples similar to your platform.
+You will need to add `seL4/libsel4/sel4_plat_include/<platform>/sel4/plat/api/constants.h`.
+Depending on your platform, this is either empty or requires defining the value for `seL4_UserTop`.
+Look in the other platform files for examples similar to your platform.
+
+You will need to add a CMake configuration file for the platform at `kernel/src/plat/<platform>/config.cmake`.
+It is best to base the contents of this file off of other platforms of the same architecture.
 
-### CMake-build system
+## Testing your seL4 port
 
-You will need to modify the build system in two places. Firstly in the kernel's platform folder, you define `kernel/src/plat/<platform>/config.cmake`, again you can look at other platforms for examples.
-Secondly, you will need to add 'KernelPlatform' to the cmake and elf-loader tools. You can do this by editing
-`tools/cmake-tool/helpers/application_settings.cmake` and `tools/elfloader-tool/CMakeLists.txt`.
+In order to sanity check your port of seL4, it is a good idea to get
+[the seL4test project](https://docs.sel4.systems/projects/sel4test/) running and passing.
```

Comment:
Could add that having a working `sel4test` patch is a requirement for contributing the platform port back to upstream seL4.

Reply:
Maybe mention then, that `sel4bench` should also work. That would ensure the libs get some fundamental porting and there is a rough picture how well the platform works.

Reply:
> Could add that having a working sel4test patch is a requirement for contributing the platform port back to upstream seL4.

Sure.

> Maybe mention then, that sel4bench should also work. That would ensure the libs get some fundamental porting and there is a rough picture how well the platform works.

I disagree. For example, sel4bench relies on getting performance data on RISC-V platforms in a way that does not work with any recent versions of OpenSBI and hence any somewhat recent RISC-V platforms also don't work (e.g the Star64).

I do not think sel4bench relies on any other libraries that are platform dependent that sel4test does not already rely on,.

Reply:
>Could add that having a working sel4test patch is a requirement for contributing the platform port back to upstream seL4.

Actually looks like I already mention that with "Once you have a working port that passes seL4test, see the guides for [contributing in general]"

or should it be clearer?

Reply:
It's more about setting up expectations. If you only read up to here then you'll think `sel4test` is optional, and we're already getting too many half-done ports that just generate more work and can't be merged.

For `sel4bench`: we do want to require that in general, because without performance data we don't know if there are problems, and `sel4bench` tends to stress test parts that `sel4test` doesn't. But it's not without issues and we need to resolve the OpenSBI problem otherwise we'll lose the ability to get useful performance data in general. So I'm fine with leaving it out of the guide here, but we need to resolve it at some point.

Reply:
> So I'm fine with leaving it out of the guide here, but we need to resolve it at some point.

Yes, of course.

> It's more about setting up expectations. If you only read up to here then you'll think sel4test is optional, and we're already getting too many half-done ports that just generate more work and can't be merged.

Makes sense, I've tried to make it clearer in https://github.com/seL4/docs/pull/233.